common/coconut-interface/src/lib.rs

```diff
@@ -0,0 +1,291 @@
+pub use coconut_rs::*;
+use getset::{CopyGetters, Getters};
+use serde::{Deserialize, Serialize};
+use url::Url;
+pub use validator_client::validator_api::Client as ValidatorAPIClient;
+use validator_client::validator_api::{
+    error::ValidatorAPIClientError, VALIDATOR_API_BLIND_SIGN, VALIDATOR_API_CACHE_VERSION,
+    VALIDATOR_API_VERIFICATION_KEY,
+};
+
+#[derive(Serialize, Deserialize, Getters, CopyGetters, Clone)]
+pub struct Credential {
+    #[getset(get = "pub")]
+    n_params: u32,
+    #[getset(get = "pub")]
+    theta: Theta,
+    public_attributes: Vec<String>,
+    #[getset(get = "pub")]
+    signature: Signature,
+}
+impl Credential {
+    pub fn new(
+        n_params: u32,
+        theta: &Theta,
+        public_attributes: &[Attribute],
+        signature: &Signature,
+    ) -> Credential {
+        Credential {
+            n_params,
+            theta: theta.clone(),
+            public_attributes: public_attributes
+                .iter()
+                .map(|attr| attr.to_bs58())
+                .collect(),
+            signature: *signature,
+        }
+    }
+
+    pub async fn init(validator_urls: Vec<String>) -> Result<Self, String> {
+        let mut state = State::init();
+        let client = ValidatorAPIClient::new();
+        let signature = get_aggregated_signature(validator_urls.clone(), &state, &client)
+            .await
+            .map_err(|e| format!("Could not aggregate signature from validators: {}", e))?;
+
+        state.signatures.push(signature);
+        let verification_key = get_aggregated_verification_key(validator_urls, &client).await?;
+        let theta = prove_credential(0, &verification_key, &state)
+            .await
+            .map_err(|e| format!("Could not prove credential: {}", e))?;
+        Ok(Credential::new(
+            state.n_attributes,
+            &theta,
+            &*state.public_attributes,
+            &signature,
+        ))
+    }
+
+    pub fn public_attributes(&self) -> Vec<Attribute> {
+        self.public_attributes
+            .iter()
+            .map(|x| Attribute::try_from_bs58(x).unwrap())
+            .collect()
+    }
+
+    pub async fn verify(&self, verification_key: &VerificationKey) -> bool {
+        let params = Parameters::new(self.n_params).unwrap();
```

Comment:
Perhaps the params should be kept on self since it's derived purely from the internal state?

Reply:
It would make more sense to do that, yes, but right now `Parameters` doesn't implement (De)Serialize , so that it can be kept inside `Credential`